src/flb_error_reporter.c

```diff
@@ -0,0 +1,225 @@
+    if (strlen(buf) > error_reporter->max_size) {
+        // truncate message
+        buf = flb_sds_copy(buf, msg, error_reporter->max_size);
+    }
+
+    message->data = flb_sds_create(buf);
+    message->len = strlen(buf);
+
+   /* clean up old message to provide enough space for new message*/
+    mk_list_foreach_safe(head, tmp, &error_reporter->messages) {
+        tmp_message = mk_list_entry(head, struct flb_error_message, _head);
+        if (error_reporter->file_size + strlen(buf) <= error_reporter->max_size) {
+            break;
+        } else {
+            error_reporter->file_size -= tmp_message->len;
+            deleted_message_count++;
+            mk_list_del(&tmp_message->_head);
+            flb_sds_destroy(tmp_message->data);
+            flb_free(tmp_message);
+        }
+    }
+    message->timestamp = time(NULL);
+
+    mk_list_add(&message->_head, &error_reporter->messages);
+
+    if (deleted_message_count == 0) {
+        f = fopen(error_reporter->file_path, "a");
+        fprintf(f, message->data);
+    } else {
+        f = fopen(error_reporter->file_path, "w");
+        mk_list_foreach_safe(head, tmp, &error_reporter->messages) {
+            tmp_message = mk_list_entry(head, struct flb_error_message, _head);
+            fprintf(f, tmp_message->data);
+        }
+    }
+    fclose(f);
+
+    error_reporter->file_size += message->len;
+    flb_sds_destroy(buf);
+
+    return 0;
+
+}
+
+/*error reporter clean up the expired message based on TTL*/
+void flb_error_reporter_clean(struct flb_error_reporter *error_reporter)
+{
+    struct mk_list *tmp;
+    struct mk_list *head;
+    struct flb_error_message *message;
+    int expired_message_count = 0;
+
+    FILE *f;
+
+    /* check the timestamp for every message and clean up expired messages*/
+    mk_list_foreach_safe(head, tmp, &error_reporter->messages) {
+        message = mk_list_entry(head, struct flb_error_message, _head);
+        if (error_reporter->ttl > time(NULL) - message->timestamp) {
+            break;
+        }
+        error_reporter->file_size -= message->len;
+        mk_list_del(&message->_head);
+        flb_sds_destroy(message->data);
+        flb_free(message);
+        expired_message_count++;
+    }
+
+    /* rewrite error report file if any message is cleaned up*/
+    if (expired_message_count > 0) {
+        f = fopen(error_reporter->file_path, "w");
+        mk_list_foreach_safe(head, tmp, &error_reporter->messages) {
+            message = mk_list_entry(head, struct flb_error_message, _head);
+            fprintf(f, message->data);
+        }
+        fclose(f);
+    }
+}
+
+/* error reporter clean up when fluent bit shutdown*/
+void flb_error_reporter_destroy(struct flb_error_reporter *error_reporter)
+{
+    if(error_reporter->file_path) {
+        flb_sds_destroy(error_reporter->file_path);
+    }
+    if (mk_list_is_empty(&error_reporter->messages) != 0) {
+        struct mk_list *tmp;
```

Comment:
the definition should be in the beginning: https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions

Reply:
updated